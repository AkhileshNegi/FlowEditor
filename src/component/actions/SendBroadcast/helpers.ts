import { BroadcastMsg, Contact, Group } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';

export const getRecipients = (action: BroadcastMsg): Asset[] => {
    const selected = action.groups.map((group: Group) => {
        return { id: group.uuid, name: group.name, type: AssetType.Group };
    });

    return selected.concat(
        action.contacts.map((contact: Contact) => {
            return { id: contact.uuid, name: contact.name, type: AssetType.Contact };
        })
    );
};
