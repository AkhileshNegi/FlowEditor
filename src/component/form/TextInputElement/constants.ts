import { CompletionOption } from '../../../store';

// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
export enum KeyValues {
    KEY_AT = '@',
    KEY_SPACE = ' ',
    KEY_ENTER = 'Enter',
    KEY_UP = 'ArrowUp',
    KEY_DOWN = 'ArrowDown',
    KEY_TAB = 'Tab',
    KEY_ESC = 'Escape',
    KEY_BACKSPACE = 'Backspace',
    KEY_N = 'n',
    KEY_P = 'p'
}

export enum TopLevelVariables {
    contact = 'contact',
    input = 'input',
    run = 'run',
    parent = 'parent',
    child = 'child',
    trigger = 'trigger'
}

export const MAX_GSM_SINGLE = 160;
export const MAX_GSM_MULTI = 153;
export const MAX_UNICODE_SINGLE = 70;
export const MAX_UNICODE_MULTI = 67;

export const COMPLETION_HELP = 'Tab to complete, enter to select';

export const getFlowOptions = (accessor?: string) => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        { name: `${prefix}.flow`, description: `The flow in which a ${accessor} run takes place` },
        {
            name: `${prefix}.flow.uuid`,
            description: `The UUID of the flow in which a ${accessor} run takes place`
        },
        {
            name: `${prefix}.flow.name`,
            description: `The name of the flow in which a ${accessor} run takes place`
        },
        {
            name: `${prefix}.flow.revision`,
            description: `The revision number of the flow in which a ${accessor} run takes place`
        }
    ];
};

export const getWebhookOptions = (accessor?: string) => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        {
            name: `${prefix}.webhook`,
            description: `The body of the response to the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.status`,
            description: `The status of the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.status_code`,
            description: `The status code returned from the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.url`,
            description: `The URL that was called by the last webhook request in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.body`,
            description: `The body of the last webhook request made in a ${accessor} run`
        },
        {
            name: `${prefix}.webhook.json`,
            description: `The JSON parsed body of the response to the last webhook request in a ${accessor} run, can access subelements`
        },
        {
            name: `${prefix}.webhook.request`,
            description: `The raw request last made in a ${accessor} run, including headers`
        },
        {
            name: `${prefix}.webhook.response`,
            description: `The raw response last received in a ${accessor} run, including headers`
        },
        { name: `${prefix}.results`, description: `Results collected in a ${accessor} run` }
    ];
};

export const getInputOptions = (accessor?: string) => {
    const prefix = accessor ? accessor : TopLevelVariables.run;
    return [
        { name: `${prefix}.input`, description: `A ${accessor} run's last input` },
        { name: `${prefix}.input.uuid`, description: `The UUID of a ${accessor} run's last input` },
        { name: `${prefix}.input.type`, description: `The type of a ${accessor} run's last input` },
        {
            name: `${prefix}.input.channel`,
            description: `The channel a ${accessor} run's last input was received on`
        },
        {
            name: `${prefix}.input.created_on`,
            description: `The time a ${accessor} run's last input was created`
        },
        {
            name: `${prefix}.input.text`,
            description: `The text of a ${accessor} run's last message`
        },
        {
            name: `${prefix}.input.attachments`,
            description: `The attachments on a ${accessor} run's last message`
        },
        {
            name: `${prefix}.input.urn`,
            description: `The URN a ${accessor} run's last input was received on`
        }
    ];
};

export const getContactOptions = (accessor?: string) => {
    const prefix = accessor ? `${accessor}.` : '';
    const descriptor = accessor
        ? accessor === TopLevelVariables.run ? `${accessor}'s` : `${accessor} run's`
        : '';
    return [
        { name: `${prefix}contact`, description: `The name of the ${descriptor} contact` },
        { name: `${prefix}contact.name`, description: `The name of the ${descriptor} contact` },
        {
            name: `${prefix}contact.first_name`,
            description: `The first name of the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.language`,
            description: `The language code for the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.fields`,
            description: `Custom fields on the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.groups`,
            description: `The groups the ${descriptor} contact is a member of`
        },
        { name: `${prefix}contact.urns`, description: `URNs on the ${descriptor} contact` },
        {
            name: `${prefix}contact.urns.tel`,
            description: `The preferred telephone number for the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.urns.telegram`,
            description: `The preferred telegram id for the ${descriptor} contact`
        },
        {
            name: `${prefix}contact.channel`,
            description: `The ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.uuid`,
            description: `The UUID of the ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.name`,
            description: `The name of the ${descriptor} contact's preferred channel`
        },
        {
            name: `${prefix}contact.channel.address`,
            description: `The address of the ${descriptor} contact's preferred channel`
        }
    ];
};

export const RUN_OPTIONS: CompletionOption[] = [
    { name: TopLevelVariables.run, description: 'A run in this flow' },
    ...getFlowOptions(),
    ...getContactOptions(TopLevelVariables.run),
    ...getInputOptions(),
    ...getWebhookOptions()
];

export const CHILD_OPTIONS: CompletionOption[] = [
    { name: TopLevelVariables.child, description: 'Run details collected in a child flow, if any' },
    ...getFlowOptions(TopLevelVariables.child),
    ...getContactOptions(TopLevelVariables.child),
    ...getInputOptions(TopLevelVariables.child),
    ...getWebhookOptions(TopLevelVariables.child)
];

export const PARENT_OPTIONS: CompletionOption[] = [
    {
        name: TopLevelVariables.parent,
        description: 'Run details collected by a parent flow, if any'
    },
    ...getFlowOptions(TopLevelVariables.parent),
    ...getContactOptions(TopLevelVariables.parent),
    ...getInputOptions(TopLevelVariables.parent),
    ...getWebhookOptions(TopLevelVariables.parent)
];

export const TRIGGER_OPTIONS: CompletionOption[] = [
    { name: 'trigger', description: 'The trigger that initiated the session' },
    { name: 'trigger.type', description: 'The type of the trigger, one of “manual” or “flow”' },
    { name: 'trigger.params', description: 'The parameters passed to the trigger' }
];

export const OPTIONS: CompletionOption[] = [
    ...getContactOptions(),
    ...RUN_OPTIONS,
    ...CHILD_OPTIONS,
    ...PARENT_OPTIONS,
    ...TRIGGER_OPTIONS
];

export const TOP_LEVEL_OPTIONS = OPTIONS.filter(
    ({ name }) =>
        name === TopLevelVariables.contact ||
        name === TopLevelVariables.input ||
        name === TopLevelVariables.run ||
        name === TopLevelVariables.parent ||
        name === TopLevelVariables.child ||
        name === TopLevelVariables.trigger
);

export const GSM: { [key: string]: number } = {
    // char: charCode
    0: 48,
    1: 49,
    2: 50,
    3: 51,
    4: 52,
    5: 53,
    6: 54,
    7: 55,
    8: 56,
    9: 57,
    '\n': 10,
    '\f': 12,
    '\r': 13,
    ' ': 32,
    '!': 33,
    '"': 34,
    '#': 35,
    $: 36,
    '%': 37,
    '&': 38,
    "'": 39,
    '(': 40,
    ')': 41,
    '*': 42,
    '+': 43,
    ',': 44,
    '-': 45,
    '.': 46,
    '/': 47,
    ':': 58,
    ';': 59,
    '<': 60,
    '=': 61,
    '>': 62,
    '?': 63,
    '@': 64,
    A: 65,
    B: 66,
    C: 67,
    D: 68,
    E: 69,
    F: 70,
    G: 71,
    H: 72,
    I: 73,
    J: 74,
    K: 75,
    L: 76,
    M: 77,
    N: 78,
    O: 79,
    P: 80,
    Q: 81,
    R: 82,
    S: 83,
    T: 84,
    U: 85,
    V: 86,
    W: 87,
    X: 88,
    Y: 89,
    Z: 90,
    '[': 91,
    '\\': 92,
    ']': 93,
    '^': 94,
    _: 95,
    a: 97,
    b: 98,
    c: 99,
    d: 100,
    e: 101,
    f: 102,
    g: 103,
    h: 104,
    i: 105,
    j: 106,
    k: 107,
    l: 108,
    m: 109,
    n: 110,
    o: 111,
    p: 112,
    q: 113,
    r: 114,
    s: 115,
    t: 116,
    u: 117,
    v: 118,
    w: 119,
    x: 120,
    y: 121,
    z: 122,
    '{': 123,
    '|': 124,
    '}': 125,
    '~': 126,
    '¡': 161,
    '£': 163,
    '¤': 164,
    '¥': 165,
    '§': 167,
    '¿': 191,
    Ä: 196,
    Å: 197,
    Æ: 198,
    Ç: 199,
    É: 201,
    Ñ: 209,
    Ö: 214,
    Ø: 216,
    Ü: 220,
    ß: 223,
    à: 224,
    ä: 228,
    å: 229,
    æ: 230,
    è: 232,
    é: 233,
    ì: 236,
    ñ: 241,
    ò: 242,
    ö: 246,
    ø: 248,
    ù: 249,
    ü: 252,
    Γ: 915,
    Δ: 916,
    Θ: 920,
    Λ: 923,
    Ξ: 926,
    Π: 928,
    Σ: 931,
    Φ: 934,
    Ψ: 936,
    Ω: 937,
    '€': 8364
};
