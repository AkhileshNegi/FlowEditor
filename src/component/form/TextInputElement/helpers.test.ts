import { FlowDefinition } from '../../../flowTypes';
import { GSM, OPTIONS } from './constants';
import { cleanMsg, filterOptions, getOptionsList, getUnicodeChars, isUnicode, pluralize } from './helpers';

const definition: FlowDefinition = require('../../../../__test__/assets/flows/a4f64f1b-85bc-477e-b706-de313a022979.json');

const optionQueryMap = OPTIONS.reduce((argMap, { name }) => {
    const lastIndex = name.lastIndexOf('.');
    if (lastIndex > -1) {
        argMap[name.slice(0, lastIndex + 2)] = true;
        argMap[name.slice(0, lastIndex + 3)] = true;
        argMap[name] = true;
    } else {
        argMap[name.slice(0, 1)] = true;
        argMap[name.slice(0, 2)] = true;
        argMap[name] = true;
        argMap[`${name}.`] = true;
    }
    return argMap;
}, {});

const winkEmoji = '😉';

describe('helpers >', () => {
    describe('isUnicode >', () => {
        it('should return true if arg is Unicode, false otherwise', () => {
            expect(isUnicode(winkEmoji)).toBeTruthy();
            Object.keys(GSM).forEach(key => expect(isUnicode(key)).toBeFalsy());
        });
    });

    describe('getUnicodeChars >', () => {
        it('should return an "empty" object if not passed a string containing Unicode chars', () => {
            expect(getUnicodeChars('abcd')).toEqual({});
        });

        it('should return a UnicodeCharMap if passed a string containing Unicode chars', () => {
            expect(getUnicodeChars(winkEmoji)).toEqual({ [winkEmoji]: true });
        });
    });

    describe('cleanMsg >', () => {
        it('should replace specified unicode characters with their GSM counterparts', () =>
            expect(cleanMsg('“”‘’— …–')).toBe(`""''- ...-`));
    });

    describe('filterOptions >', () => {
        it('should return an empty array if not passed a query', () =>
            expect(filterOptions(OPTIONS)).toEqual([]));

        Object.keys(optionQueryMap).forEach(query =>
            it(`should filter options for "${query}"`, () =>
                expect(filterOptions(OPTIONS, query)).toMatchSnapshot())
        );
    });

    describe('pluralize', () => {
        it("should apply suffix to noun if count param isn't 1", () => {
            expect(pluralize(2, 'character')).toBe('characters');
        });

        it("shouldn't apply suffix to noun if count is 1", () => {
            expect(pluralize(1, 'character')).toBe('character');
        });
    });

    describe('getOptionsList', () => {
        it('should include only our predefined options if autocomplete arg is falsy', () => {
            expect(getOptionsList(false, {}).length).toBe(OPTIONS.length);
            expect(getOptionsList(false, {})).toMatchSnapshot();
        });

        it('should include result names if autocomplete arg is truthy', () => {
            const resultsCompletionMap = {
                'ecc70717-dd25-4795-8dc2-0361265a1e29': {
                    name: '@run.results.color',
                    description: 'Result for "color"'
                },
                'aa1bdfea-d319-45a3-b08a-55a0485e4306': {
                    name: '@run.results.color',
                    description: 'Result for "color"'
                }
            };
            const optionsList = getOptionsList(true, resultsCompletionMap);
            // Expect duplicate results to be removed.
            const expectedLength = OPTIONS.length + 1;

            expect(optionsList.length).toBe(expectedLength);
            expect(optionsList).toMatchSnapshot();
        });
    });
});
