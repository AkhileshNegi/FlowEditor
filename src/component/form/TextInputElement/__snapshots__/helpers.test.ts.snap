// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers > filterOptions > should filter options for "c" 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact",
  },
  Object {
    "description": "Run details after running a child flow",
    "name": "child",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "ch" 1`] = `
Array [
  Object {
    "description": "Run details after running a child flow",
    "name": "child",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "child" 1`] = `
Array [
  Object {
    "description": "Run details after running a child flow",
    "name": "child",
  },
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "child." 1`] = `
Array [
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "child.r" 1`] = `
Array [
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "child.re" 1`] = `
Array [
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "child.results" 1`] = `
Array [
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "co" 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact" 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact",
  },
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact." 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.f" 1`] = `
Array [
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.fi" 1`] = `
Array [
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.fields" 1`] = `
Array [
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.g" 1`] = `
Array [
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.gr" 1`] = `
Array [
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.groups" 1`] = `
Array [
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.l" 1`] = `
Array [
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.la" 1`] = `
Array [
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.language" 1`] = `
Array [
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.n" 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.na" 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.name" 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.u" 1`] = `
Array [
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.ur" 1`] = `
Array [
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.urns" 1`] = `
Array [
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
  Object {
    "description": "The preferred telephone number for the contact.",
    "name": "contact.urns.tel",
  },
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.urns.t" 1`] = `
Array [
  Object {
    "description": "The preferred telephone number for the contact.",
    "name": "contact.urns.tel",
  },
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.urns.te" 1`] = `
Array [
  Object {
    "description": "The preferred telephone number for the contact.",
    "name": "contact.urns.tel",
  },
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.urns.tel" 1`] = `
Array [
  Object {
    "description": "The preferred telephone number for the contact.",
    "name": "contact.urns.tel",
  },
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "contact.urns.telegram" 1`] = `
Array [
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "i" 1`] = `
Array [
  Object {
    "description": "The last input from the contact if any.",
    "name": "input",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "in" 1`] = `
Array [
  Object {
    "description": "The last input from the contact if any.",
    "name": "input",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "input" 1`] = `
Array [
  Object {
    "description": "The last input from the contact if any.",
    "name": "input",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "input." 1`] = `Array []`;

exports[`helpers > filterOptions > should filter options for "p" 1`] = `
Array [
  Object {
    "description": "Run details if being called from a parent flow",
    "name": "parent",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "pa" 1`] = `
Array [
  Object {
    "description": "Run details if being called from a parent flow",
    "name": "parent",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "parent" 1`] = `
Array [
  Object {
    "description": "Run details if being called from a parent flow",
    "name": "parent",
  },
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "parent." 1`] = `
Array [
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "parent.r" 1`] = `
Array [
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "parent.re" 1`] = `
Array [
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "parent.results" 1`] = `
Array [
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "r" 1`] = `
Array [
  Object {
    "description": "Run details",
    "name": "run",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "ru" 1`] = `
Array [
  Object {
    "description": "Run details",
    "name": "run",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run" 1`] = `
Array [
  Object {
    "description": "Run details",
    "name": "run",
  },
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run." 1`] = `
Array [
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run.c" 1`] = `
Array [
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run.co" 1`] = `
Array [
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run.contact" 1`] = `
Array [
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run.r" 1`] = `
Array [
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run.re" 1`] = `
Array [
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "run.results" 1`] = `
Array [
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "w" 1`] = `
Array [
  Object {
    "description": "The body of the webhook response",
    "name": "webhook",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "we" 1`] = `
Array [
  Object {
    "description": "The body of the webhook response",
    "name": "webhook",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook" 1`] = `
Array [
  Object {
    "description": "The body of the webhook response",
    "name": "webhook",
  },
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook." 1`] = `
Array [
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.b" 1`] = `
Array [
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.bo" 1`] = `
Array [
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.body" 1`] = `
Array [
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.j" 1`] = `
Array [
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.js" 1`] = `
Array [
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.json" 1`] = `
Array [
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.r" 1`] = `
Array [
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.re" 1`] = `
Array [
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.request" 1`] = `
Array [
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.response" 1`] = `
Array [
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.s" 1`] = `
Array [
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.st" 1`] = `
Array [
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.status" 1`] = `
Array [
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.status_code" 1`] = `
Array [
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.u" 1`] = `
Array [
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.ur" 1`] = `
Array [
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
]
`;

exports[`helpers > filterOptions > should filter options for "webhook.url" 1`] = `
Array [
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
]
`;

exports[`helpers > getOptionsList should include only our predefined options if autocomplete arg is falsy 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact",
  },
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
  Object {
    "description": "The preferred telephone number for the contact.",
    "name": "contact.urns.tel",
  },
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
  Object {
    "description": "The last input from the contact if any.",
    "name": "input",
  },
  Object {
    "description": "Run details",
    "name": "run",
  },
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
  Object {
    "description": "Run details after running a child flow",
    "name": "child",
  },
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
  Object {
    "description": "Run details if being called from a parent flow",
    "name": "parent",
  },
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
  Object {
    "description": "The body of the webhook response",
    "name": "webhook",
  },
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
]
`;

exports[`helpers > getOptionsList should include result names if autocomplete arg is truthy 1`] = `
Array [
  Object {
    "description": "The name of the contact.",
    "name": "contact",
  },
  Object {
    "description": "The name of the contact.",
    "name": "contact.name",
  },
  Object {
    "description": "The language code for the contact.",
    "name": "contact.language",
  },
  Object {
    "description": "Custom fields on the contact.",
    "name": "contact.fields",
  },
  Object {
    "description": "The groups for the contact.",
    "name": "contact.groups",
  },
  Object {
    "description": "URNs on the contact.",
    "name": "contact.urns",
  },
  Object {
    "description": "The preferred telephone number for the contact.",
    "name": "contact.urns.tel",
  },
  Object {
    "description": "The preferred telegram id for the contact.",
    "name": "contact.urns.telegram",
  },
  Object {
    "description": "The last input from the contact if any.",
    "name": "input",
  },
  Object {
    "description": "Run details",
    "name": "run",
  },
  Object {
    "description": "The contact in this run",
    "name": "run.contact",
  },
  Object {
    "description": "Results for the run",
    "name": "run.results",
  },
  Object {
    "description": "Run details after running a child flow",
    "name": "child",
  },
  Object {
    "description": "The results for the child flow",
    "name": "child.results",
  },
  Object {
    "description": "Run details if being called from a parent flow",
    "name": "parent",
  },
  Object {
    "description": "The results for the parent flow",
    "name": "parent.results",
  },
  Object {
    "description": "The body of the webhook response",
    "name": "webhook",
  },
  Object {
    "description": "The status of the webhook call",
    "name": "webhook.status",
  },
  Object {
    "description": "The status code returned from the webhook",
    "name": "webhook.status_code",
  },
  Object {
    "description": "The URL which was called",
    "name": "webhook.url",
  },
  Object {
    "description": "The body of the webhook response",
    "name": "webhook.body",
  },
  Object {
    "description": "The JSON parsed body of the response, can access subelements",
    "name": "webhook.json",
  },
  Object {
    "description": "The raw request of the webhook including headers",
    "name": "webhook.request",
  },
  Object {
    "description": "The raw response of the webhook including headers",
    "name": "webhook.response",
  },
  Object {
    "description": "Result for \\"color\\"",
    "name": "run.results.color",
  },
]
`;
